airflow/ti_deps/dep_context.py

```diff
@@ -0,0 +1,125 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from airflow.ti_deps.deps.dag_ti_slots_available_dep import DagTISlotsAvailableDep
+from airflow.ti_deps.deps.dag_unpaused_dep import DagUnpausedDep
+from airflow.ti_deps.deps.dagrun_exists_dep import DagrunExistsDep
+from airflow.ti_deps.deps.exec_date_after_start_date_dep import ExecDateAfterStartDateDep
+from airflow.ti_deps.deps.not_running_dep import NotRunningDep
+from airflow.ti_deps.deps.not_skipped_dep import NotSkippedDep
+from airflow.ti_deps.deps.pool_has_space_dep import PoolHasSpaceDep
+from airflow.ti_deps.deps.runnable_exec_date_dep import RunnableExecDateDep
+from airflow.ti_deps.deps.valid_state_dep import ValidStateDep
+from airflow.utils.state import State
+
+
+class DepContext(object):
+    """
+    A base class for contexts that specifies which dependencies should be evaluated in
+    the context for a task instance to satisfy the requirements of the context. Also
+    stores state related to the context that can be used by dependendency classes.
+
+    For example there could be a SomeRunContext that subclasses this class which has
+    dependencies for:
+    - Making sure there are slots available on the infrastructure to run the task instance
+    - A task-instance's task-specific dependencies are met (e.g. the previous task
+      instance completed successfully)
+    - ...
+
+    :param deps: The context-specific dependencies that need to be evaluated for a
+        task instance to run in this execution context.
+    :type deps: set(BaseTIDep)
+    :param flag_upstream_failed: This is a hack to generate the upstream_failed state
+        creation while checking to see whether the task instance is runnable. It was the
+        shortest path to add the feature. This is bad since this class should be pure (no
+        side effects).
+    :type flag_upstream_failed: boolean
+    :param ignore_all_deps: Whether or not the context should ignore all ignoreable
+        dependencies. Overrides the other ignore_* parameters
+    :type ignore_all_deps: boolean
+    :param ignore_depends_on_past: Ignore depends_on_past parameter of DAGs (e.g. for
+        Backfills)
+    :type ignore_depends_on_past: boolean
+    :param ignore_task_deps: Ignore task-specific dependencies such as depends_on_past and
+        trigger rule
+    :type ignore_task_deps: boolean
+    :param ignore_ti_state: Ignore the task instance's previous failure/success
+    :type ignore_ti_state: boolean
+    """
+    def __init__(
+            self,
+            deps=None,
```

Comment:
At first I thought that it's a bit of a crossover referencing deps in here, but now that I think about it makes a lot of sense to directly whitelist and/or blacklist deps in the context object. 

On a sidenote, I just though that `start_date` as in the start_date of a backfill mutating where "depends_on_past" get skipped (overriding the DAG's start_date), may find its place nicely in this object.